### PR TITLE
feat: add graphql operations for comments

### LINKED ISSUE #164: feat: add graphql operations for comments
queries 
* Get comments by article id
* Get all comments 

mutation 
* Create comment
* Delete comment by id

### PR DESCRIPTION
**Pull request by Shurtu-gal**
### Solved or Linked Issues

Reference any and all issues solved and related to this pull request here. Mention if there are none.
Fixed #164 

### Description and Changes

Please describe your pull request with all changes made.

### Your checklist for this pull request

🚨Please review the [guidelines for contributing](../README.md) to this repository.

- [x] Make sure you are requesting to **pull a work/topic/feature/bugfix branch** (right side). Don't request your default branch!
- [x] Make sure you are making a pull request against the **development** (left side). Also you should start _your branch_ off _default branch_ and follow the branch naming guidelines.
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.
- [x] I have added necessary documentation and mentioned the related issues (if appropriate)
- [x] I have added the appropriate labels, requested a review from a repository manager and assigned myself to this pull request.

### Post merge checklist

- [x] Follow steps from the [guidelines for contributing](../blob/development/CONTRIBUTING.md) to this repository.

### CODE REVIEW

File: server/schema/comment/comment.model.js [hunk 1]
```diff
@@ -9,7 +9,7 @@
  * @since 0.1.0
  */
 
-const { Schema, model } = require('mongoose');
+const { Schema, model, Model: _Model } = require('mongoose');
```
Comment (120EE0692) [inline server/schema/comment/comment.model.js:12]: @Shurtu-gal why this change was done?

File: server/schema/comment/comment.mutation.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+const CommentType = require('./comment.type');
+
+module.exports = new GraphQLObjectType({
+  name: 'CommentMutation',
+  fields: {
+    createComment: {
+      description: 'Create a comment',
+      type: CommentType,
+      args: {
+        content: { type: new GraphQLNonNull(new GraphQLList(GraphQLJSONObject)) },
```
Comment (120EE0692) [inline server/schema/comment/comment.mutation.js:34]: @Shurtu-gal why the content type is like this, why not simply a string?

File: server/schema/comment/comment.resolver.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    try {
+      const _count = await Comment.countNumberOfComments(id, parentType);
+      return _count;
+    } catch (error) {
+      throw APIError(null, error);
+    }
+  },
+  createComment: async (
+    _parent,
+    { authorID, parentID, parentType },
```
Comment (120EE0692) [inline server/schema/comment/comment.resolver.js:59]: @Shurtu-gal without any comment text?

File: server/schema/comment/comment.model.js (hunk 1 above)
Comment (Shurtu-gal) [reply server/schema/comment/comment.model.js:12]: This is to get the required intellisense using JSDoc during work.

File: server/schema/comment/comment.mutation.js (hunk 2 above)
Comment (Shurtu-gal) [reply server/schema/comment/comment.mutation.js:34]: Will make it a string. This was for styling in the comment.

File: server/schema/comment/comment.mutation.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        authorID: { type: new GraphQLNonNull(GraphQLID) },
+        parentID: { type: new GraphQLNonNull(GraphQLID) },
+        parentType: { type: new GraphQLNonNull(GraphQLString) },
```
Comment (120EE0692) [inline server/schema/comment/comment.mutation.js:37]: @Shurtu-gal create a enum type for this

File: server/schema/comment/comment.datasources.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+    const _comment = await CommentModel.findByIdAndUpdate(
+      id,
+      {
+        content,
+        updatedBy: UserSession.valid(session, authToken) ? mid : null,
+      },
+      { new: true, session: mdbSession }
+    );
+    await mdbSession.commitTransaction();
```
Comment (DesignrKnight) [inline server/schema/comment/comment.datasources.js:81]: Again, why transaction for single write operation?

File: server/schema/comment/comment.datasources.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+const countNumberOfComments = (parentID, parentModel) =>
+  CommentModel.countDocuments({
+    'parent.reference': parentID,
+    'parent.model': parentModel,
+  });
+
+const create = async (authorID, content, parentID, parentType, session, authToken, mid) => {
+  const mdbSession = await connection.startSession();
+  try {
+    mdbSession.startTransaction();
```
Comment (120EE0692) [inline server/schema/comment/comment.datasources.js:36]: @Shurtu-gal remove this as well.

File: server/schema/comment/comment.datasources.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            reference: parentID,
+            model: parentType,
+          },
+          createdBy: UserSession.valid(session, authToken) ? mid : null,
+        },
+      ],
+      { session: mdbSession }
+    );
+    await mdbSession.commitTransaction();
+    await mdbSession.endSession();
```
Comment (120EE0692) [inline server/schema/comment/comment.datasources.js:57]: @Shurtu-gal remove this as well.

File: server/schema/comment/comment.datasources.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    return _comment;
+  } catch (error) {
+    await mdbSession.abortTransaction();
+    await mdbSession.endSession();
```
Comment (120EE0692) [inline server/schema/comment/comment.datasources.js:62]: @Shurtu-gal remove this as well.

**Review by 120EE0692 — CHANGES_REQUESTED** [review]

File: server/schema/comment/comment.query.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+const { getListOfComments, getCommentById, countOfComments } = require('./comment.resolver');
+
+const CommentType = require('./comment.type');
+
+module.exports = new GraphQLObjectType({
+  name: 'CommentQuery',
+  fields: {
+    getListOfComments: {
+      description: 'Retrieves comments for given list of ids (default all) in descending order of creation time',
+      type: new GraphQLList(new GraphQLNonNull(CommentType)),
```
Comment (120EE0692) [inline server/schema/comment/comment.query.js:32]: @Shurtu-gal will this query work, when we have no comments.
What is the reason of adding GraphQLNonNull here?

**Review by 120EE0692 — CHANGES_REQUESTED** [review]

File: server/schema/comment/comment.query.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+      args: {
+        ids: {
+          description: 'List of Ids of comments to be retrieved',
+          type: new GraphQLList(new GraphQLNonNull(GraphQLID)),
```
Comment (120EE0692) [inline server/schema/comment/comment.query.js:36]: @Shurtu-gal GraphQLNonNull should be avoided here the query work fine even if we don't pass the list of ids.

File: server/schema/comment/comment.query.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      args: {
+        id: {
+          description: 'Id of article or comment',
+          type: GraphQLID,
+        },
+        parentType: {
+          description: 'Type of parent',
+          type: GraphQLString,
+        },
+      },
```
Comment (120EE0692) [inline server/schema/comment/comment.query.js:72]: @Shurtu-gal add GraphQLNonNull here because we can't get the required result with out the parent id and type.

File: server/schema/comment/comment.resolver.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+      if (
+        !UserPermission.exists(session, authToken, decodedToken, 'comment.write.new') &&
+        !UserPermission.exists(session, authToken, decodedToken, 'comment.write.all')
+      ) {
+        throw APIError('FORBIDDEN', null, 'User does not have required permission to create comment');
+      }
+
+      const _comment = await Comment.create(authorID, content, parentID, parentType, session, authToken, mid);
+
+      return _comment[0];
```
Comment (120EE0692) [inline server/schema/comment/comment.resolver.js:72]: @Shurtu-gal check the return type, if it an array then destructure it there.

i.e [_comment] = await Comment.create()

File: server/schema/comment/comment.resolver.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      throw APIError(null, error);
+    }
+  },
+  updateCommentContent: async (
+    _parent,
+    { id, content, authorID },
+    { session, authToken, decodedToken, mid, API: { Comment } }
+  ) => {
+    try {
+      await canMutateComment(session, authToken, decodedToken, id, authorID, mid, Comment);
```
Comment (120EE0692) [inline server/schema/comment/comment.resolver.js:97]: @Shurtu-gal make sure only user with role comment.write.all should able to perform this operation.

**Review by 120EE0692 — CHANGES_REQUESTED** [review]

File: server/schema/comment/comment.query.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+          type: GraphQLInt,
+        },
+        offset: {
+          description: 'No of Comments to be skipped | pagination',
+          type: GraphQLInt,
+        },
+      },
+      resolve: getListOfComments,
+    },
+    getCommentById: {
```
Comment (120EE0692) [inline server/schema/comment/comment.query.js:49]: @Shurtu-gal is this returning parent id ?

File: server/schema/comment/comment.query.js [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+  // GraphQLJSONObject,
+} = require('../scalars');
+const { getListOfComments, getCommentById, countOfComments } = require('./comment.resolver');
+
+const CommentType = require('./comment.type');
+
+module.exports = new GraphQLObjectType({
+  name: 'CommentQuery',
+  fields: {
+    getListOfComments: {
```
Comment (120EE0692) [inline server/schema/comment/comment.query.js:30]: @Shurtu-gal is this returning root article id ?

File: server/schema/comment/comment.datasources.js [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+        updatedBy: UserSession.valid(session, authToken) ? mid : null,
+      },
+      { new: true }
+    );
+
+    return _comment;
+  } catch (error) {
+    throw APIError(null, error);
+  }
+};
```
Comment (120EE0692) [inline server/schema/comment/comment.datasources.js:100]: @Shurtu-gal in which situation we will use this?

Comment (120EE0692) [inline server/schema/comment/comment.datasources.js:100]: @Shurtu-gal in which situation we will use this?

**Review by 120EE0692 — APPROVED** [review]

**Comment by 120EE0692** [comment]
| Role Name          | comment.read.public | comment.read.unapproved | comment.write.new | comment.write.approved | comment.write.self | comment.write.delete | comment.approve.all |
| ------------------ | ------------------- | ----------------------- | ----------------- | ---------------------- | ------------------ | -------------------- | ------------------- |
| comment.basic      | true                | false                   | true              | false                  | true               | false                | false               |
| comment.verified   | true                | false                   | true              | true                   | true               | false                | false               |
| comment.admin      | true                | true                    | true              | true                   | true               | false                | true                |
| comment.superadmin | true                | true                    | true              | true                   | true               | true                 | true                |


@Shurtu-gal can you please verify this permission check are implemented or not.

**Comment by rutajdash** [comment]
@120EE0692 close this PR if inactive

**Comment by Shurtu-gal** [comment]
> Role Name	comment.read.public	comment.read.unapproved	comment.write.new	comment.write.approved	comment.write.self	comment.write.delete	comment.approve.all
> comment.basic	true	false	true	false	true	false	false
> comment.verified	true	false	true	true	true	false	false
> comment.admin	true	true	true	true	true	false	true
> comment.superadmin	true	true	true	true	true	true	true
> @Shurtu-gal can you please verify this permission check are implemented or not.

I will do it by EOD today. I was out for some days

**Comment by Shurtu-gal** [comment]
@120EE0692 Kindly Review this.

**Review by 120EE0692 — APPROVED** [review]

**Comment by 120EE0692** [comment]
@Shurtu-gal please rebase this with the main branch.

**Comment by Shurtu-gal** [comment]
@120EE0692 Done

**Review by 120EE0692 — APPROVED** [review]
